File: backend/src/routes/exchangeRates.js
import express from 'express';
import { query } from '../db.js';
import { authMiddleware } from '../middleware/auth.js';

const router = express.Router();

// All routes require authentication
router.use(authMiddleware);

// Middleware to check if user is admin
const isAdmin = async (req, res, next) => {
  const result = await query('SELECT role FROM users WHERE id = $1', [req.user.id]);
  if (result.rows[0]?.role !== 'admin') {
    return res.status(403).json({ error: 'Admin access required' });
  }
  next();
};

// Get all exchange rates (public endpoint)
router.get('/', async (req, res) => {
  try {
    const result = await query(`
      SELECT er.*, u.email as updated_by_email 
      FROM exchange_rates er
      LEFT JOIN users u ON er.updated_by = u.id
      ORDER BY er.currency
    `);
    res.json(result.rows);
  } catch (error) {
    console.error('Error fetching exchange rates:', error);
    res.status(500).json({ error: 'Failed to fetch exchange rates' });
  }
});

// Update exchange rate (admin only)
router.put('/:currency', isAdmin, async (req, res) => {
  try {
    const { currency } = req.params;
    const { rate } = req.body;

    if (!rate || isNaN(rate) || rate <= 0) {
      return res.status(400).json({ error: 'Invalid rate value' });
    }

    const result = await query(
      `UPDATE exchange_rates 
       SET rate = $1, updated_at = CURRENT_TIMESTAMP, updated_by = $2 
       WHERE currency = $3 
       RETURNING *`,
      [rate, req.user.id, currency.toUpperCase()]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Currency not found' });
    }

    res.json(result.rows[0]);
  } catch (error) {
    console.error('Error updating exchange rate:', error);
    res.status(500).json({ error: 'Failed to update exchange rate' });
  }
});

// Add new currency (admin only)
router.post('/', isAdmin, async (req, res) => {
  try {
    const { currency, rate } = req.body;

    if (!currency || currency.length !== 3) {
      return res.status(400).json({ error: 'Currency must be a 3-letter code' });
    }

    if (!rate || isNaN(rate) || rate <= 0) {
      return res.status(400).json({ error: 'Invalid rate value' });
    }

    const result = await query(
      `INSERT INTO exchange_rates (currency, rate, updated_by) 
       VALUES ($1, $2, $3) 
       RETURNING *`,
      [currency.toUpperCase(), rate, req.user.id]
    );

    res.status(201).json(result.rows[0]);
  } catch (error) {
    if (error.code === '23505') { // Unique violation
      return res.status(409).json({ error: 'Currency already exists' });
    }
    console.error('Error creating exchange rate:', error);
    res.status(500).json({ error: 'Failed to create exchange rate' });
  }
});

export default router;

File: fastapi_app/routes/exchange_rates.py
"""
Exchange rates routes.
"""
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from models import ExchangeRateCreate, ExchangeRateUpdate, ExchangeRateResponse, MessageResponse
from auth import get_current_user, require_admin, User
from database import query, query_one, execute

router = APIRouter(prefix="/exchange-rates", tags=["Exchange Rates"])


@router.get("", response_model=List[ExchangeRateResponse])
async def list_exchange_rates(current_user: User = Depends(get_current_user)):
    """
    List all exchange rates.

    Exchange rates are relative to RON (Romanian Leu).
    For example, if EUR rate is 4.97, then 1 EUR = 4.97 RON.
    """
    rates = query(
        """
        SELECT id, currency, rate, updated_at, updated_by
        FROM exchange_rates
        ORDER BY currency
        """
    )
    return [ExchangeRateResponse(**r) for r in rates]


@router.post("", response_model=ExchangeRateResponse, status_code=status.HTTP_201_CREATED)
async def create_exchange_rate(
    rate: ExchangeRateCreate,
    admin: User = Depends(require_admin)
):
    """
    Add a new currency exchange rate. Admin only.

    - **currency**: 3-letter currency code (e.g., 'EUR', 'USD', 'GBP')
    - **rate**: Exchange rate relative to RON (e.g., 4.97 for EUR)
    """
    existing = query_one(
        "SELECT id FROM exchange_rates WHERE currency = %s",
        (rate.currency.upper(),)
    )

    if existing:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Exchange rate for '{rate.currency}' already exists. Use PUT to update."
        )

    result = execute(
        """
        INSERT INTO exchange_rates (currency, rate, updated_by)
        VALUES (%s, %s, %s)
        RETURNING id, currency, rate, updated_at, updated_by
        """,
        (rate.currency.upper(), rate.rate, admin.id)
    )

    return ExchangeRateResponse(**result)


@router.put("/{currency}", response_model=ExchangeRateResponse)
async def update_exchange_rate(
    currency: str,
    rate: ExchangeRateUpdate,
    admin: User = Depends(require_admin)
):
    """
    Update an exchange rate. Admin only.

    - **currency**: 3-letter currency code
    - **rate**: New exchange rate relative to RON
    """
    result = execute(
        """
        UPDATE exchange_rates
        SET rate = %s, updated_at = NOW(), updated_by = %s
        WHERE currency = %s
        RETURNING id, currency, rate, updated_at, updated_by
        """,
        (rate.rate, admin.id, currency.upper())
    )

    if not result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Exchange rate for '{currency}' not found"
        )

    return ExchangeRateResponse(**result)


@router.delete("/{currency}", response_model=MessageResponse)
async def delete_exchange_rate(
    currency: str,
    admin: User = Depends(require_admin)
):
    """
    Delete an exchange rate. Admin only.
    """
    result = execute(
        "DELETE FROM exchange_rates WHERE currency = %s RETURNING id",
        (currency.upper(),)
    )

    if not result:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Exchange rate for '{currency}' not found"
        )

    return MessageResponse(message=f"Exchange rate for '{currency}' deleted successfully")